File: sdks/java/basys.sdk/backends/org/eclipse/basyx/aas/backend/connected/ConnectedOperation.java
package org.eclipse.basyx.aas.backend.connected;

import java.util.List;
import java.util.function.Function;

import org.eclipse.basyx.aas.api.resources.IOperation;
import org.eclipse.basyx.aas.metamodel.hashmap.aas.submodelelement.operation.OperationVariable;
import org.eclipse.basyx.vab.core.proxy.VABElementProxy;

public class ConnectedOperation extends ConnectedElement implements IOperation {

	public ConnectedOperation(String path, VABElementProxy proxy) {
		super(path, proxy);
	}

	@Override
	public Object invoke(Object... params) throws Exception {
		// FIXME: endpoint contains a string and invokable is not explicitly in the
		// meta model
		return getProxy().invoke(constructPath("invokable"), params);
	}

	@SuppressWarnings("unchecked")
	@Override
	public List<OperationVariable> getParameterTypes() {
		return (List<OperationVariable>) getProxy().readElementValue("in");
	}

	@SuppressWarnings("unchecked")
	@Override
	public List<OperationVariable> getReturnTypes() {
		return (List<OperationVariable>) getProxy().readElementValue("out");
	}

	// TODO Based on the requirement the following functions need to be implemented later
	@Override
	public void SetParameterTypes(List<OperationVariable> in) {
		// TODO Auto-generated method stub
		
	}

	@Override
	public void setReturnTypes(List<OperationVariable> out) {
		// TODO Auto-generated method stub
		
	}

	@Override
	public void setInvocable(Function<Object[], Object[]> endpoint) {
		// TODO Auto-generated method stub
		
	}

	@Override
	public Function<Object[], Object[]> getInvocable() {
		// TODO Auto-generated method stub
		return null;
	}

}

File: sdks/c++/basys.sdk.cc/src/aas/backend/connected/aas/submodelelement/operation/ConnectedOperation.cpp
/*
 * ConnectedOperation.cpp
 *
 *      Author: wendel
 */

#include "ConnectedOperation.h"

namespace basyx {
namespace aas {
namespace backend {
namespace connected {

ConnectedOperation::ConnectedOperation(std::shared_ptr<vab::core::proxy::IVABElementProxy> proxy) :
  ConnectedSubmodelElement(proxy)
{}

void ConnectedOperation::setId(const std::string & id)
{
  this->setIdWithLocalCheck(id);
}

std::string ConnectedOperation::getId() const
{
  return this->getIdWithLocalCheck();
}

operation_var_list ConnectedOperation::getParameterTypes() const
{
  // todo
  return operation_var_list(); // this->getProxyCollection(std::string(submodelelement::operation::OperationPaths::INPUT));
}

operation_var_list ConnectedOperation::getReturnTypes() const
{
  // todo
  return operation_var_list(); // this->getProxyCollection(std::string(submodelelement::operation::OperationPaths::INPUT));
}

basyx::any ConnectedOperation::invoke(basyx::objectCollection_t & parameters) const
{
  return this->getProxy()->invoke("", parameters);

}

std::shared_ptr<basyx::function_base> ConnectedOperation::getInvocable() const
{
  // Not supported on remote side.
  return nullptr;
}

}
}
}
}

